### PR TITLE
[O365] | Migrate to CEL input

### PR DESCRIPTION
**Pull request by kcreddy**
## What does this PR do?
Adds CEL input for fetching events from O365 Management API. Deprecates the existing o365audit input.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ]

## How to test this PR locally
- Run `elastic-package stack down && elastic-package build && elastic-package stack up -d -v && eval "$(elastic-package stack shellinit)" && elastic-package test system --generate  -v`
Produces:
```
.......
2023/07/12 13:01:10 DEBUG found 0 hits in logs-o365.audit-ep data stream: index_not_found_exception: no such index [logs-o365.audit-ep] Status=404
2023/07/12 13:01:11 DEBUG found 0 hits in logs-o365.audit-ep data stream: index_not_found_exception: no such index [logs-o365.audit-ep] Status=404
2023/07/12 13:01:12 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:01:13 DEBUG found 27 hits in logs-o365.audit-ep data stream
........
........
2023/07/12 13:02:10 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:11 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:12 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:13 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:14 DEBUG found 6 hits in logs-o365.audit-ep data stream
.........
.........
--- Test results for package: o365 - START ---
╭─────────┬─────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├─────────┼─────────────┼───────────┼───────────┼────────┼───────────────┤
│ o365    │ audit       │ system    │ o365audit │ PASS   │ 27.106824625s │
│ o365    │ audit       │ system    │ cel       │ PASS   │ 26.534162792s │
╰─────────┴─────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: o365 - END   ---
Done
```

## Related issues

- Closes https://github.com/elastic/integrations/issues/5045

### CODE REVIEW

**Comment by jamiehynds** [comment]
@kcreddy can you confirm the behaviour when a user of the current O365 integration updates to the new version that uses CEL? Will the switch to CEL be automatic without any user intervention or will httpjson be disabled and a user has re-enter their O365 API manually?

**Comment by kcreddy** [comment]
> Will the switch to CEL be automatic without any user intervention or will httpjson be disabled and a user has re-enter their O365 API manually?

Users would have to setup the new CEL input manually. The input parameters are pretty similar to what they have been using with O365 Audit input ie., `Azure Client ID, Tenant ID, Secret`, etc,. The package and datastream the events end up is also same. i.e., `data_stream.dataset: o365.audit`. In other words we are not creating a new package or a datastream.

The screenshot of setup looks like this:
Screenshot 2023-06-21 at 10 33 29 AM
Screenshot 2023-06-21 at 10 33 44 AM
------------------------------------------------------------------------------


> can you confirm the behaviour when a user of the current O365 integration updates to the new version that uses CEL?

After getting suggestions from @andrewkroh, the behaviour would result in some duplicate events for users. That is because when the users upgrade the integration, the index rollover happens. So, even the existing `fingerprint` processor inside ingest pipeline on event's `Id` field would still create duplicates for last few days (up until 7 days back). We have the leverage to decide how far back we want to start pulling the events. We could also give the users ability to choose their start time. [Reference](https://learn.microsoft.com/en-us/office/office-365-management-api/office-365-management-activity-api-reference#working-with-the-office-365-management-activity-api:~:text=Both%20must%20be%20specified%20(or%20both%20omitted)%20and%20they%20must%20be%20no%20more%20than%2024%20hours%20apart%2C%20with%20the%20start%20time%20no%20more%20than%207%20days%20in%20the%20past.%20By%20default%2C%20if%20startTime%20and%20endTime%20are%20omitted%2C%20then%20the%20content%20available%20in%20the%20last%2024%20hours%20is%20returned.) 
The behaviour still needs to be tested though.

**Comment by jamiehynds** [comment]
@kcreddy is the user prompted to input the credentials once they've upgraded the integration? We need to minimise the chances of a user upgrade the package, saving it and not realising that ingestion has stopped until they enter their Client ID & Secret. If we enable the datastream by default, I presume they won't be able to save the integration and will have to populate the details before continuing? We should also make this obvious in the changelog too. 

I'd suggest adjusting the description of the datastreams too:
Old datastream using O365 input: "Collect Office 365 Audit Logs - Deprecated"
New datastream using CEL: "Collect Office 365 Audit Logs via Graph API" (assuming we're using the Graph API?)

**Comment by kcreddy** [comment]
Hey @jamiehynds,
Below is how the upgrade scenario is going to look like for the user:

https://github.com/elastic/integrations/assets/11301409/e188905b-4da5-4699-b3fd-96f34cbe74e1

Basically, the user will have to upgrade the integration policy too once the integration is upgraded to latest. Once that integration policy is updated, and since the new CEL input is enabled by default, the user will have to provide details for the CEL input to start. I have updated the title and description in the previous o365audit input to `Deprecated`. I also updated the `README` docs with the steps to follow when upgrading the integration from `< 1.7.0` to `>= 1.7.0`. Let me know if anything is missing or needs to be added/changed.

Also, during the tests, I noted that duplicates are not getting created. So, we are good there.

File: packages/o365/manifest.yml [hunk 1]
```diff
@@ -29,7 +29,10 @@ policy_templates:
     description: Collect logs from Office 365
     inputs:
       - type: o365audit
-        title: "Collect Office 365 audit logs"
-        description: "Collect audit logs from Office 365 via the Management Activity API"
+        title: "Collect Office 365 audit logs - Deprecated. Please use the CEL input instead."
```
Comment (andrewkroh) [inline packages/o365/manifest.yml:32]:
```suggestion
        title: "Collect Office 365 audit logs - Deprecated. Please disable this and use the CEL input instead."
```

I was thinking the message somewhere should tell them to disable it explicitly.

Comment (kcreddy) [reply packages/o365/manifest.yml]: Updated the manifest as suggested.

**Comment by jamiehynds** [comment]
@kcreddy thanks for putting the video together - really helpful. 

When a user upgrades the the integration the 'upgrade integration policies' option is selected, but in this case the policy isn't updated. It's probably a bit misleading, but not sure we can solve for that. 

After upgrading the integration, a user has to go to integration policies to enable the new input and then upgrade the policy. I worry that users may not inattentively know to go to integration policies after the integration upgrade. 

I don't think people will know what CEL is. Could we adjust the wording of the new datastream to 'Collect Office 365 Audit Logs via Graph API' (or whatever API we're hitting)?

Am I correct in saying v1.17 of the integration will not include the deprecated input? So for new users using the O365 integration for the first time, will only have the option to use the CEL input, or will they also see the deprecated input too?

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+{{#if enable_request_tracer}}
+resource.tracer.filename: "../../logs/httpjson/http-request-trace-*.ndjson"
+{{/if}}
+
+state:
+  want_more: false
+  base:
+    list_contents_url: "{{url}}/api/v1.0/{{azure_tenant_id}}/activity/feed/subscriptions/content?contentType="
+    list_contents_start_time: "{{initial_interval}}"
+
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:60]: Add a
```
redact:
  fields: ~
```
stanza. We will log a warning if there is no redact config even if it is not required for the specific integration so that when it is required it's less likely to be missed.

Are there any fields in state that may be sensitive? (The auth secrets won't be in state, so you are fine there).

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                                            + "&startTime=" + (content_type_state_created_at.parse_time(time_layout.RFC3339) + duration("1s")).format(time_layout.RFC3339) 
+                                            + "&endTime=" + ((content_type_state_created_at.parse_time(time_layout.RFC3339) + duration("24h")).as(calc_end_time, 
+                                                calc_end_time <= now() ? calc_end_time : now())).format(time_layout.RFC3339) ) ))
+                            :
+                                // initial run when no cursor state exists i.e., polling from initial_interval
+                                state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}"  
+                                    + "&startTime=" + (now() - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
+                                    + "&endTime=" + ((now() - duration(state.base.list_contents_start_time) + duration("24h")).as(calc_end_time, 
+                                        calc_end_time <= now() ? calc_end_time : now())).format(time_layout.RFC3339)
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:87]:
```suggestion
                                        calc_end_time <= now ? calc_end_time : now)).format(time_layout.RFC3339)
```

Is there a reason to use `now()` over `now`? The difference between the two is that (static) `now` is the time the CEL evaluation started and `now()` is the time that the `now()` call was made. Generally you will want the static case.

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                                                    "content_type": content_type,
+                                                    "content_created_at": { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max(),
+                                                    "next_page": (has(list_contents_resp.Header) && has(list_contents_resp.Header.NextPageUri) ) ? list_contents_resp.Header.NextPageUri[0] : (has(list_contents_resp.Header) && has(list_contents_resp.Header.Nextpageuri)) ? list_contents_resp.Header.Nextpageuri[0] : "",
+                                                    // keep fetching more if (nextpageuri exists) or (max time returned date != today's date)
+                                                    "want_more_content": ( (has(list_contents_resp.Header) && (has(list_contents_resp.Header.NextPageUri) && (list_contents_resp.Header.NextPageUri.size() > 0)) || (has(list_contents_resp.Header.Nextpageuri) && (list_contents_resp.Header.Nextpageuri.size() > 0)) ) || ( { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max().split('T')[0] != now().format(time_layout.RFC3339).split('T')[0] ) ) ? true : false
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:105]:
```suggestion
                                                    "want_more_content": ( (has(list_contents_resp.Header) && (has(list_contents_resp.Header.NextPageUri) && (list_contents_resp.Header.NextPageUri.size() > 0)) || (has(list_contents_resp.Header.Nextpageuri) && (list_contents_resp.Header.Nextpageuri.size() > 0)) ) || ( { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max().split('T')[0] != now.format(time_layout.RFC3339).split('T')[0] ) )
```

Recommend using a custom time format string without the zone if you don't need it, `s/time_layout.RFC3339/"2006-01-02"/g` (or when we are building on a later version of Go and mito is updated, `s/time_layout.RFC3339/time_layout.DateOnly/g`).

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                                                    list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().endTime[0]
+                                                :
+                                                    list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().startTime[0]
+                                        ,
+                                        "next_page": "",
+                                        "want_more_content": (
+                                            has(list_contents_resp.StatusCode) 
+                                            && list_contents_resp.StatusCode == 200 
+                                            && (list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().endTime[0].split('T')[0] != now().format(time_layout.RFC3339).split('T')[0]) 
+                                            ) ? true : false
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:125]:
```suggestion
                                            )
```

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                (
+                    {
+                        "events_per_content_type": [],
+                        "content_type": content_type,
+                        "content_created_at": 
+                            ( has (state.cursor) && has(state.cursor.content_types_state_as_list)) ?
+                                state.cursor.content_types_state_as_list.filter(e, e.content_type == content_type)[0].content_created_at
+                            :
+                                (
+                                    (now() - duration("168h")).format(time_layout.RFC3339)
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:141]:
```suggestion
                                    (now - duration("168h")).format(time_layout.RFC3339)
```

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+program: |
+  "{{content_types}}".split(",").map(content_type_raw, content_type_raw.trim_space()).as(content_types, 
+    content_types.map(content_type,
+        (request("POST", "{{url}}/api/v1.0/{{azure_tenant_id}}/activity/feed/subscriptions/start?contentType=" + content_type + "&PublisherIdentifier={{azure_tenant_id}}"))
+        .do_request().as(start_subs_resp,
+            bytes(start_subs_resp.Body).decode_json().as(start_subs_resp_body, 
+                (has(start_subs_resp_body.status) && start_subs_resp_body.status == 'enabled') || (has(start_subs_resp_body.error) && has(start_subs_resp_body.error.code) && start_subs_resp_body.error.code == "AF20024") ?
+                    // When start-subscription API returns success or if already started subscription,
+                    (request("GET", 
+                        ( has(state.want_more) && state.want_more && has(state.cursor) && has(state.cursor.content_types_state_as_list) && (state.cursor.content_types_state_as_list.filter(e, e.content_type == content_type)[0].next_page != "") ) ?
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:70]: Since you define `state.want_more: false` in the initial state, you don't need the `has(state.want_more)` here.

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+                                    .as(content_type_state, content_type_state[0].content_created_at.as(content_type_state_created_at, 
+                                        (state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}" 
+                                            + "&startTime=" + (content_type_state_created_at.parse_time(time_layout.RFC3339) + duration("1s")).format(time_layout.RFC3339) 
+                                            + "&endTime=" + ((content_type_state_created_at.parse_time(time_layout.RFC3339) + duration("24h")).as(calc_end_time, 
+                                                calc_end_time <= now() ? calc_end_time : now())).format(time_layout.RFC3339) ) ))
+                            :
+                                // initial run when no cursor state exists i.e., polling from initial_interval
+                                state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}"  
+                                    + "&startTime=" + (now() - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339)
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:85]:
```suggestion
                                    + "&startTime=" + (now - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
```

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                                        (state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}" 
+                                            + "&startTime=" + (content_type_state_created_at.parse_time(time_layout.RFC3339) + duration("1s")).format(time_layout.RFC3339) 
+                                            + "&endTime=" + ((content_type_state_created_at.parse_time(time_layout.RFC3339) + duration("24h")).as(calc_end_time, 
+                                                calc_end_time <= now() ? calc_end_time : now())).format(time_layout.RFC3339) ) ))
+                            :
+                                // initial run when no cursor state exists i.e., polling from initial_interval
+                                state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}"  
+                                    + "&startTime=" + (now() - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
+                                    + "&endTime=" + ((now() - duration(state.base.list_contents_start_time) + duration("24h")).as(calc_end_time,
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:86]:
```suggestion
                                    + "&endTime=" + ((now - duration(state.base.list_contents_start_time) + duration("24h")).as(calc_end_time, 
```

File: packages/o365/data_stream/audit/elasticsearch/ingest_pipeline/default.yml [hunk 10]
```diff
@@ -24,10 +28,21 @@ processors:
...
       ignore_missing: true
+  - convert:
+      field: o365audit.ListBaseType
+      type: string
+      tag: convert-listbasetype-to-string
+      if: ctx.o365audit?.ListBaseType != null
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag fail-{{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
Comment (efd6) [inline packages/o365/data_stream/audit/elasticsearch/ingest_pipeline/default.yml:44]:
```suggestion
            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```

**Comment by jamiehynds** [comment]
> Hey @jamiehynds, Below is how the upgrade scenario is going to look like for the user:
> 
>  O365-upgrade.1.mp4 
> Basically, the user will have to upgrade the integration policy too once the integration is upgraded to latest. Once that integration policy is updated, and since the new CEL input is enabled by default, the user will have to provide details for the CEL input to start. I have updated the title and description in the previous o365audit input to `Deprecated`. I also updated the `README` docs with the steps to follow when upgrading the integration from `< 1.17.0` to `>= 1.17.0`. Let me know if anything is missing or needs to be added/changed.
> 
> Also, during the tests, I noted that duplicates are not getting created. So, we are good there.
> 
> I didn't use the Graph API, since not all of the Office 365 and Azure AD activity events are available under Graph API. We are still using the [Office 365 Management Activity API](https://learn.microsoft.com/en-us/office/office-365-management-api/office-365-management-activity-api-reference) which is still actively supported.

@amitkanfer not sure if there's any improvements planned on the Fleet side, but this is a good example of a case where we have to make a signifiant change to an integration (changing an input) and difficult to inform users on the impact of the change. It's also a challenge to guide users on the steps to follow to ensure ingestion continues post-upgrade. In this case they have to disable the old data stream and re-enter their creds in a new data stream. While we try to keep these types of changes to a minimum, we find ourselves making them incases where a vendor deprecates an API or we have to switch inputs within packages. No action needed now, just an FYI.

**Comment by kcreddy** [comment]
Hey @jamiehynds 
Update - the version is now `1.18.0`. So, the old o265audit input is now at `1.17.0`

> When a user upgrades the the integration the 'upgrade integration policies' option is selected, but in this case the policy isn't updated. It's probably a bit misleading, but not sure we can solve for that.
After upgrading the integration, a user has to go to integration policies to enable the new input and then upgrade the policy. I worry that users may not inattentively know to go to integration policies after the integration upgrade.

That is in fact correct and is misleading the users to think that upgrade is successful without actually upgrading the integration policy. But I am not sure what we can do here rather than providing a README documentation. Maybe the Fleet UI could redirect the user automatically to `integration policies` tab when there is such `Review field conflicts` warning comes up.

> I don't think people will know what CEL is. Could we adjust the wording of the new datastream to 'Collect Office 365 Audit Logs via Graph API' (or whatever API we're hitting)?

I updated the manifest.yml title to `Collect Office 365 audit logs via the Management Activity API using CEL Input`

> Am I correct in saying v1.17 of the integration will not include the deprecated input? So for new users using the O365 integration for the first time, will only have the option to use the CEL input, or will they also see the deprecated input too?

New version `v1.18` includes the deprecated input as well, but is disabled by default. So, new users also see the deprecated input as well. The CEL input is enabled by default. So, they need to configure the required parameters to save and make it run.

-----------

Also, I have now recorded a new video where I totally removed the deprecated input o365audit from the manifest.yml. Even now the users will have to manually upgrade the integration policy after upgrading the integration. But inside the policy UI, there is no deprecated input. It is bit cleaner, but the drawback could be that if something went wrong and users would like to continue reusing the old input, they would need to downgrade the integration version.

Here's the video of how it would look like without the deprecated input. Please let me know which of these you prefer, if the new integration version needs to contain or not contain the old deprecated input.

https://github.com/elastic/integrations/assets/11301409/fa1c4d35-de5f-4170-9b33-2d18d88191bb

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 2 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:60]: Apart from the `client_secret` which is used for oauth, there are not other sensitive fields. I added the stanza 👍🏼

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 3 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Got it. `now` makes sense. Updated the code to replace `now()` with `now`

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 5 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Updated as suggested

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 6 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Updated to `now`

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 7 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Makes sense 👍🏼 Updated as suggested

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 8 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Updated to `now`

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 9 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Updated to `now`

File: packages/o365/data_stream/audit/elasticsearch/ingest_pipeline/default.yml (hunk 10 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/elasticsearch/ingest_pipeline/default.yml]: Updated as suggested

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 4 above)
Comment (efd6) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: It just [calls `time.Time.Format`, and takes a string](https://github.com/elastic/mito/blob/3a86f09b8d5f3872229eb66129a863222d2924e6/lib/time.go#L201-L211). Those globals are just strings.

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 11]
```diff
@@ -138,7 +141,7 @@ program: |
                                 state.cursor.content_types_state_as_list.filter(e, e.content_type == content_type)[0].content_created_at
                             :
                                 (
-                                    (now() - duration("168h")).format(time_layout.RFC3339)
+                                    (now - duration("168h")).format(time_layout.RFC3339)
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:144]: Do we want to have the "168h" parameterisable?

Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: True. Updated it to value containing initial interval

**Comment by jamiehynds** [comment]
> Also, I have now recorded a new video where I totally removed the deprecated input o365audit from the manifest.yml. Even now the users will have to manually upgrade the integration policy after upgrading the integration. But inside the policy UI, there is no deprecated input. It is bit cleaner, but the drawback could be that if something went wrong and users would like to continue reusing the old input, they would need to downgrade the integration version.
> 
> Here's the video of how it would look like without the deprecated input. Please let me know which of these you prefer, if the new integration version needs to contain or not contain the old deprecated input.

@kcreddy while removing the deprecated input is certainly cleaner, I think given the popularity of the integration and the extent the the httpjson to CEL change, it makes more sense to have an escape hatch available, at least for a release or two as users transition to the upgraded integration. So I'd vote to keeping the deprecated input as fallback for now. WDYT @andrewkroh?

**Comment by kcreddy** [comment]
> If the deprecated input remains for a release or two, then reverting to httpjson would be a matter or enabling httpjson and disabling cel via the existing toggle. The user would not need to enter any new config because the old config would be saved. Correct?

The old input for this package is `o365audit`. But yes, I verified that the old config is being saved. Hence the user can just disabled `CEL` input and enable the old `o365audit` input.

> If the httpjson input was removed, what's the level of difficulty in downgrading the package version if something goes wrong?

That would mean they would need to uninstall the integration and reinstall the older version. I found this tracking issue: https://github.com/elastic/kibana/issues/131801. Currently there does not seem to be an easier way to downgrade the package versions in Fleet.

**Comment by andrewkroh** [comment]
Because downgrading is complicated and the user would have lost their existing configuration for the o365audit input, I think keeping the input around for a brief transition period will help if there is some unknown issue.

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 3 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs]: Hey @efd6 I reverted this change back to `now()` because `now` is being evaluated to the time when CEL evaluation initially started, remains the same after each poll. Not sure if its a bug or its meant to work that way.  

Example errors  when I used `now` (consecutive polls i.e., after `3 min`):
`Start time and end time must both be specified (or both omitted) and must be less than or equal to 24 hours apart, with the start time prior to  end time and start time no more than 7 days in the past. StartTime:2023-07-18T09:39:16Z, EndTime:2023-07-18T09:39:14Z`

`Start time and end time must both be specified (or both omitted) and must be less than or equal to 24 hours apart, with the start time prior to end time and start time no more than 7 days in the past. StartTime:2023-07-18T09:39:17Z, EndTime:2023-07-18T09:39:14Z`

Notice that the `endTime` which is evaluated to `now` remains the same.

Comment (efd6) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:90]: Hmmmm. That's problematic. I will look into it. I'm surprised, and it's in language, and so non-trivial. I would suggest a work-around of adding this prefix to the program and using `state.now`.
```
state.with({"now": now()}).
```

Comment (efd6) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:90]: Fix at #36107. In the mean time, I suggest using the workaround above.

Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:90]: Thanks Dan!

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+                                                    "content_type": content_type,
+                                                    "content_created_at": { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max(),
+                                                    "next_page": (has(list_contents_resp.Header) && has(list_contents_resp.Header.NextPageUri) ) ? list_contents_resp.Header.NextPageUri[0] : (has(list_contents_resp.Header) && has(list_contents_resp.Header.Nextpageuri)) ? list_contents_resp.Header.Nextpageuri[0] : "",
+                                                    // keep fetching more if (nextpageuri exists) or (max time returned date != today's date)
+                                                    "want_more_content": ( (has(list_contents_resp.Header) && (has(list_contents_resp.Header.NextPageUri) && (list_contents_resp.Header.NextPageUri.size() > 0)) || (has(list_contents_resp.Header.Nextpageuri) && (list_contents_resp.Header.Nextpageuri.size() > 0)) ) || ( { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max().split('T')[0] != now().format("2006-01-02") ) )
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:108]: This can be made less expensive. The important thing here is that `||` and `&&` are not short circuiting, so the collation is always evaluated, even when the first or second operands of the first two disjunctions are true. Normally, I don't worry about this, but in this case, `collate` is reasonably expensive and so we should avoid it if possible.

There's notes about this [here](https://github.com/google/cel-spec/blob/master/doc/langdef.md#logical-operators).

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                                                (has(list_contents_resp.StatusCode) && list_contents_resp.StatusCode == 200 ) ?
+                                                    list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().endTime[0]
+                                                :
+                                                    list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().startTime[0]
+                                        ,
+                                        "next_page": "",
+                                        "want_more_content": (
+                                            has(list_contents_resp.StatusCode) 
+                                            && list_contents_resp.StatusCode == 200 
+                                            && (list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().endTime[0].split('T')[0] != now().format("2006-01-02"))
```
Comment (efd6) [inline packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:127]: We can do a similar thing here.

**Review by efd6 — APPROVED** [review]
Minor issues only.

The CEL code issues can be a future optimisation is you prefer.

File: packages/o365/data_stream/audit/agent/stream/cel.yml.hbs (hunk 12 above)
Comment (kcreddy) [reply packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:108]: Thanks Dan for the suggestions. I will definitely take it as an enhancement.
